File: plugin/src/knowledge/providers/shodan.ts
import { requestUrl } from "obsidian";
import type { KnowledgeProvider, LookupOptions, LookupResult, KnowledgeEntry } from "../types";

const SHODAN_API = "https://api.shodan.io";

interface ShodanHostResult {
  ip_str?: string;
  hostnames?: string[];
  org?: string;
  isp?: string;
  asn?: string;
  country_name?: string;
  city?: string;
  os?: string;
  ports?: number[];
  vulns?: string[];
  tags?: string[];
  data?: Array<{
    port: number;
    transport: string;
    product?: string;
    version?: string;
    banner?: string;
  }>;
}

interface ShodanSearchResult {
  matches?: Array<{
    ip_str: string;
    port: number;
    org?: string;
    hostnames?: string[];
    product?: string;
    os?: string;
    country_name?: string;
    asn?: string;
  }>;
  total?: number;
}

interface ShodanDnsResult {
  [domain: string]: {
    type: string;
    value: string;
    subdomain?: string;
  }[];
}

export interface ShodanConfig {
  apiKey?: string;
}

export class ShodanProvider implements KnowledgeProvider {
  readonly name = "shodan";
  private apiKey: string;

  constructor(config?: ShodanConfig) {
    this.apiKey = config?.apiKey ?? "";
  }

  configure(config: ShodanConfig): void {
    if (config.apiKey !== undefined) this.apiKey = config.apiKey;
  }

  async isAvailable(): Promise<boolean> {
    return this.apiKey.length > 0;
  }

  async lookup(query: string, options?: LookupOptions): Promise<LookupResult> {
    const maxResults = options?.maxResults ?? 5;

    if (!this.apiKey) {
      return {
        success: false,
        provider: this.name,
        entries: [],
        error: "Shodan API key not configured",
      };
    }

    try {
      // Check if query is an IP address
      if (this.isIpAddress(query)) {
        const entry = await this.lookupHost(query);
        return {
          success: true,
          provider: this.name,
          entries: entry ? [entry] : [],
        };
      }

      // Check if query is a domain
      if (this.isDomain(query)) {
        const entries = await this.lookupDns(query);
        return {
          success: true,
          provider: this.name,
          entries: entries.slice(0, maxResults),
        };
      }

      // Otherwise, search
      const entries = await this.search(query, maxResults);
      return {
        success: true,
        provider: this.name,
        entries,
      };
    } catch (error) {
      const message = error instanceof Error ? error.message : "Unknown error";
      return {
        success: false,
        provider: this.name,
        entries: [],
        error: `Shodan lookup failed: ${message}`,
      };
    }
  }

  private isIpAddress(query: string): boolean {
    return /^(\d{1,3}\.){3}\d{1,3}$/.test(query);
  }

  private isDomain(query: string): boolean {
    return /^[a-zA-Z0-9][a-zA-Z0-9-]*\.[a-zA-Z]{2,}$/.test(query);
  }

  private async lookupHost(ip: string): Promise<KnowledgeEntry | null> {
    const response = await requestUrl({
      url: `${SHODAN_API}/shodan/host/${ip}?key=${this.apiKey}`,
      method: "GET",
      throw: false,
    });

    if (response.status === 404) {
      return null;
    }

    if (response.status !== 200) {
      throw new Error(`Host lookup failed: ${response.status}`);
    }

    const data = response.json as ShodanHostResult;

    const hostnames = data.hostnames?.join(", ") || "No hostnames";
    const location = [data.city, data.country_name].filter(Boolean).join(", ");
    const ports = data.ports?.slice(0, 10).join(", ") || "None";

    const lines: string[] = [];
    lines.push(`IP: ${data.ip_str}`);
    if (data.org) lines.push(`Organization: ${data.org}`);
    if (location) lines.push(`Location: ${location}`);
    lines.push(`Open Ports: ${ports}`);
    if (data.os) lines.push(`OS: ${data.os}`);
    if (data.vulns && data.vulns.length > 0) {
      lines.push(`Vulnerabilities: ${data.vulns.slice(0, 5).join(", ")}`);
    }

    return {
      title: `${data.ip_str} (${hostnames})`,
      summary: lines.join("\n"),
      url: `https://www.shodan.io/host/${ip}`,
      source: "shodan",
      metadata: {
        ip: data.ip_str,
        hostnames: data.hostnames,
        org: data.org,
        isp: data.isp,
        asn: data.asn,
        country: data.country_name,
        city: data.city,
        os: data.os,
        ports: data.ports,
        vulns: data.vulns,
        tags: data.tags,
      },
    };
  }

  private async lookupDns(domain: string): Promise<KnowledgeEntry[]> {
    const response = await requestUrl({
      url: `${SHODAN_API}/dns/domain/${domain}?key=${this.apiKey}`,
      method: "GET",
      throw: false,
    });

    if (response.status !== 200) {
      throw new Error(`DNS lookup failed: ${response.status}`);
    }

    const data = response.json as ShodanDnsResult;
    const entries: KnowledgeEntry[] = [];

    // Return domain info as a single entry
    const records: string[] = [];
    for (const [subdomain, recordList] of Object.entries(data)) {
      if (Array.isArray(recordList)) {
        for (const record of recordList.slice(0, 5)) {
          records.push(`${subdomain}: ${record.type} -> ${record.value}`);
        }
      }
    }

    if (records.length > 0) {
      entries.push({
        title: domain,
        summary: `DNS Records:\n${records.slice(0, 10).join("\n")}`,
        url: `https://www.shodan.io/domain/${domain}`,
        source: "shodan",
        metadata: {
          domain,
          recordCount: records.length,
        },
      });
    }

    return entries;
  }

  private async search(query: string, limit: number): Promise<KnowledgeEntry[]> {
    const params = new URLSearchParams({
      key: this.apiKey,
      query,
    });

    const response = await requestUrl({
      url: `${SHODAN_API}/shodan/host/search?${params}`,
      method: "GET",
      throw: false,
    });

    if (response.status !== 200) {
      throw new Error(`Search failed: ${response.status}`);
    }

    const data = response.json as ShodanSearchResult;

    if (!data.matches) {
      return [];
    }

    return data.matches.slice(0, limit).map((match) => {
      const hostnames = match.hostnames?.join(", ") || "No hostnames";
      const lines: string[] = [];
      lines.push(`Port: ${match.port}`);
      if (match.org) lines.push(`Org: ${match.org}`);
      if (match.product) lines.push(`Product: ${match.product}`);
      if (match.country_name) lines.push(`Country: ${match.country_name}`);

      return {
        title: `${match.ip_str}:${match.port} (${hostnames})`,
        summary: lines.join(" | "),
        url: `https://www.shodan.io/host/${match.ip_str}`,
        source: "shodan",
        metadata: {
          ip: match.ip_str,
          port: match.port,
          org: match.org,
          hostnames: match.hostnames,
          product: match.product,
          os: match.os,
          country: match.country_name,
          asn: match.asn,
        },
      };
    });
  }
}

File: crates/mcp/src/knowledge/shodan.rs
use reqwest::blocking::Client;
use serde::Deserialize;
use std::collections::HashMap;

use super::{KnowledgeEntry, KnowledgeProvider, LookupOptions, LookupResult};

const SHODAN_API: &str = "https://api.shodan.io";

pub struct ShodanProvider {
    client: Client,
    api_key: String,
}

#[derive(Debug, Deserialize)]
#[allow(dead_code)]
struct HostResult {
    ip_str: Option<String>,
    hostnames: Option<Vec<String>>,
    org: Option<String>,
    isp: Option<String>,
    asn: Option<String>,
    country_name: Option<String>,
    city: Option<String>,
    os: Option<String>,
    ports: Option<Vec<u16>>,
    vulns: Option<Vec<String>>,
    tags: Option<Vec<String>>,
}

#[derive(Debug, Deserialize)]
#[allow(dead_code)]
struct SearchResult {
    matches: Option<Vec<SearchMatch>>,
    total: Option<i64>,
}

#[derive(Debug, Deserialize)]
#[allow(dead_code)]
struct SearchMatch {
    ip_str: String,
    port: u16,
    org: Option<String>,
    hostnames: Option<Vec<String>>,
    product: Option<String>,
    os: Option<String>,
    country_name: Option<String>,
    asn: Option<String>,
}

impl ShodanProvider {
    pub fn new() -> Self {
        Self {
            client: Client::builder()
                .user_agent("vault-tree-mcp/0.1 (https://github.com/Hydepwns/vault-tree)")
                .build()
                .unwrap_or_else(|_| Client::new()),
            api_key: String::new(),
        }
    }

    pub fn with_api_key(api_key: impl Into<String>) -> Self {
        Self {
            client: Client::builder()
                .user_agent("vault-tree-mcp/0.1 (https://github.com/Hydepwns/vault-tree)")
                .build()
                .unwrap_or_else(|_| Client::new()),
            api_key: api_key.into(),
        }
    }

    fn is_ip_address(query: &str) -> bool {
        query.split('.').count() == 4
            && query.split('.').all(|part| part.parse::<u8>().is_ok())
    }

    fn lookup_host(&self, ip: &str) -> Result<Option<KnowledgeEntry>, String> {
        let url = format!("{}/shodan/host/{}?key={}", SHODAN_API, ip, self.api_key);

        let response = self.client.get(&url).send().map_err(|e| e.to_string())?;

        if response.status().as_u16() == 404 {
            return Ok(None);
        }

        if !response.status().is_success() {
            return Err(format!("host lookup failed: {}", response.status()));
        }

        let data: HostResult = response.json().map_err(|e| e.to_string())?;

        let ip_str = data.ip_str.as_deref().unwrap_or(ip);
        let hostnames = data.hostnames.as_ref()
            .map(|h| h.join(", "))
            .filter(|s| !s.is_empty())
            .unwrap_or_else(|| "No hostnames".to_string());

        let location: Vec<&str> = [data.city.as_deref(), data.country_name.as_deref()]
            .into_iter()
            .flatten()
            .collect();

        let ports = data.ports.as_ref()
            .map(|p| p.iter().take(10).map(|n| n.to_string()).collect::<Vec<_>>().join(", "))
            .unwrap_or_else(|| "None".to_string());

        let mut lines = Vec::new();
        lines.push(format!("IP: {}", ip_str));
        if let Some(org) = &data.org {
            lines.push(format!("Organization: {}", org));
        }
        if !location.is_empty() {
            lines.push(format!("Location: {}", location.join(", ")));
        }
        lines.push(format!("Open Ports: {}", ports));
        if let Some(os) = &data.os {
            lines.push(format!("OS: {}", os));
        }
        if let Some(vulns) = &data.vulns {
            if !vulns.is_empty() {
                lines.push(format!("Vulnerabilities: {}", vulns.iter().take(5).cloned().collect::<Vec<_>>().join(", ")));
            }
        }

        let mut metadata = HashMap::new();
        metadata.insert("ip".to_string(), serde_json::json!(ip_str));
        if let Some(h) = &data.hostnames {
            metadata.insert("hostnames".to_string(), serde_json::json!(h));
        }
        if let Some(org) = &data.org {
            metadata.insert("org".to_string(), serde_json::json!(org));
        }
        if let Some(p) = &data.ports {
            metadata.insert("ports".to_string(), serde_json::json!(p));
        }
        if let Some(v) = &data.vulns {
            metadata.insert("vulns".to_string(), serde_json::json!(v));
        }

        Ok(Some(KnowledgeEntry {
            title: format!("{} ({})", ip_str, hostnames),
            summary: lines.join("\n"),
            url: Some(format!("https://www.shodan.io/host/{}", ip)),
            source: "shodan".to_string(),
            metadata: Some(metadata),
        }))
    }

    fn search(&self, query: &str, limit: usize) -> Result<Vec<KnowledgeEntry>, String> {
        let url = format!(
            "{}/shodan/host/search?key={}&query={}",
            SHODAN_API,
            self.api_key,
            urlencoding::encode(query)
        );

        let response = self.client.get(&url).send().map_err(|e| e.to_string())?;

        if !response.status().is_success() {
            return Err(format!("search failed: {}", response.status()));
        }

        let data: SearchResult = response.json().map_err(|e| e.to_string())?;

        Ok(data
            .matches
            .unwrap_or_default()
            .into_iter()
            .take(limit)
            .map(|m| {
                let hostnames = m.hostnames.as_ref()
                    .map(|h| h.join(", "))
                    .filter(|s| !s.is_empty())
                    .unwrap_or_else(|| "No hostnames".to_string());

                let mut parts = Vec::new();
                parts.push(format!("Port: {}", m.port));
                if let Some(org) = &m.org {
                    parts.push(format!("Org: {}", org));
                }
                if let Some(product) = &m.product {
                    parts.push(format!("Product: {}", product));
                }
                if let Some(country) = &m.country_name {
                    parts.push(format!("Country: {}", country));
                }

                let mut metadata = HashMap::new();
                metadata.insert("ip".to_string(), serde_json::json!(m.ip_str));
                metadata.insert("port".to_string(), serde_json::json!(m.port));
                if let Some(org) = &m.org {
                    metadata.insert("org".to_string(), serde_json::json!(org));
                }
                if let Some(h) = &m.hostnames {
                    metadata.insert("hostnames".to_string(), serde_json::json!(h));
                }

                KnowledgeEntry {
                    title: format!("{}:{} ({})", m.ip_str, m.port, hostnames),
                    summary: parts.join(" | "),
                    url: Some(format!("https://www.shodan.io/host/{}", m.ip_str)),
                    source: "shodan".to_string(),
                    metadata: Some(metadata),
                }
            })
            .collect())
    }
}

impl Default for ShodanProvider {
    fn default() -> Self {
        Self::new()
    }
}

impl KnowledgeProvider for ShodanProvider {
    fn name(&self) -> &'static str {
        "shodan"
    }

    fn is_available(&self) -> bool {
        !self.api_key.is_empty()
    }

    fn lookup(&self, query: &str, options: &LookupOptions) -> LookupResult {
        let limit = options.max_results.unwrap_or(5);

        if self.api_key.is_empty() {
            return LookupResult::error(self.name(), "Shodan API key not configured");
        }

        // Check if query is an IP address
        if Self::is_ip_address(query) {
            match self.lookup_host(query) {
                Ok(Some(entry)) => return LookupResult::success(self.name(), vec![entry]),
                Ok(None) => return LookupResult::success(self.name(), vec![]),
                Err(e) => return LookupResult::error(self.name(), e),
            }
        }

        // Search
        match self.search(query, limit) {
            Ok(entries) => LookupResult::success(self.name(), entries),
            Err(e) => LookupResult::error(self.name(), e),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn shodan_requires_api_key() {
        let provider = ShodanProvider::new();
        assert!(!provider.is_available());
    }
}
